pkg/apis/batch/validation/validation.go

```diff
@@ -296,9 +300,13 @@ func ValidateCronJob(cronJob *batch.CronJob, opts apivalidation.PodValidationOpt
 func ValidateCronJobUpdate(job, oldJob *batch.CronJob, opts apivalidation.PodValidationOptions) field.ErrorList {
 	allErrs := apivalidation.ValidateObjectMetaUpdate(&job.ObjectMeta, &oldJob.ObjectMeta, field.NewPath("metadata"))
 	allErrs = append(allErrs, ValidateCronJobSpec(&job.Spec, field.NewPath("spec"), opts)...)
+	allUpdateErrs := ValidateCronJobSpecUpdate(&job.Spec, &oldJob.Spec, field.NewPath("spec"), opts)
+	if len(allUpdateErrs) == 0 {
+		allErrs = apivalidation.RemoveFromErrorList(allErrs, "setting pvc resources limits is not supported")
```

Comment:
This RemoveFromErrorList idea has a really bad smell to me.

First off, we shouldn't be using different strings, they should be set using a constant to prevent typos. Then the fact that this constant needs to be passed across different packages is already a danger sign---it means we're coupling batch validation, for example, tightly to the core validation (in that it needs to know the exact error messages). This is going to be unmaintainable long term.

The other is that it's making this subtle implicit assumption that anything in ValidateCronJobSpec that causes the limits error, is going to be checked identically in ValidateCronJobSpecUpdate. I think that's true now, but is going to be hard to maintain, particular because that assumption is satisfied in ValidateVolumesUpdate, which is far from where the error is actually generated in ValidatePersistentVolumeClaimSpecUpdate. Too much nonlocality. We're just creating landmines for future generations to step on.

Ok, so what's the alternative? 

I think what's typically done elsewhere is that a deep copy of the new resources is made, all resource limits are cleared out for volumes where the old resource had limits, and then Validate would be called. That is, instead of assuming any errors seen in Validate are spurious if Validate*Update passes, the resource is explicitly cleaned. That works well because it's super clear at every point what's going on, there aren't dependencies between different checks in different parts of the code.

The problem of course is that the resource limits are buried really far down. But it's always a PodTemplateSpec or something like that?

What if we used a pattern like

```
func ValidateXUpdate(newX, oldX) {
  xCopy = DeepCopy(...)
  CleanPodTemplate(xCopy.path.to.PodTemplateSpec)
  errs = append(errs, ValidateX(xCopy))
  errs = append(errs, ValidatePodTemplateUpdate(newX.path.to.PodTemplateSpec, oldX.path.to.PodTemplateSpec)
}
```

Then this would also cut down the towers of ValidateCronJobUpdate -> ValidateCronJobSpecUpdate -> ValidateJobTemplateSpecUpdate -> etc etc that you've had to do. 

Possibly there will have to be a couple of these things, CleanPodTemplateSpec, CleanVolumeSourceList, I'm not sure, you'll have to go through how all this is done.

Reply:
Yeah, I fully agree that it has a lot of problem from maintainability perspective.

Appreciate for the kind and detailed instructions. I have fixed them all. Thanks.